File: app/src/main/java/eu/kanade/mangafeed/ui/download/DownloadHolder.java
package eu.kanade.mangafeed.ui.download;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import butterknife.Bind;
import butterknife.ButterKnife;
import eu.kanade.mangafeed.R;
import eu.kanade.mangafeed.data.download.model.Download;

public class DownloadHolder extends RecyclerView.ViewHolder {

    @Bind(R.id.download_title) TextView downloadTitle;
    @Bind(R.id.download_progress) ProgressBar downloadProgress;
    @Bind(R.id.download_progress_text) TextView downloadProgressText;

    public DownloadHolder(View view) {
        super(view);
        ButterKnife.bind(this, view);
    }

    public void onSetValues(Download download) {
        downloadTitle.setText(download.chapter.name);

        if (download.pages == null) {
            downloadProgress.setProgress(0);
            downloadProgress.setMax(1);
            downloadProgressText.setText("");
        } else {
            downloadProgress.setMax(download.pages.size() * 100);
            setDownloadProgress(download);
            setDownloadedPages(download);
        }
    }

    public void setDownloadedPages(Download download) {
        String progressText = download.downloadedImages + "/" + download.pages.size();
        downloadProgressText.setText(progressText);
    }

    public void setDownloadProgress(Download download) {
        if (downloadProgress.getMax() == 1)
            downloadProgress.setMax(download.pages.size() * 100);
        downloadProgress.setProgress(download.totalProgress);
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/download/DownloadHolder.kt
package eu.kanade.tachiyomi.ui.download

import android.view.View
import androidx.recyclerview.widget.ItemTouchHelper
import eu.davidea.viewholders.FlexibleViewHolder
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.download.model.Download
import eu.kanade.tachiyomi.databinding.DownloadItemBinding
import eu.kanade.tachiyomi.util.view.popupMenu

/**
 * Class used to hold the data of a download.
 * All the elements from the layout file "download_item" are available in this class.
 *
 * @param view the inflated view for this holder.
 * @constructor creates a new download holder.
 */
class DownloadHolder(private val view: View, val adapter: DownloadAdapter) :
    FlexibleViewHolder(view, adapter) {

    private val binding = DownloadItemBinding.bind(view)

    init {
        setDragHandleView(binding.reorder)
        binding.menu.setOnClickListener { it.post { showPopupMenu(it) } }
    }

    private lateinit var download: Download

    /**
     * Binds this holder with the given category.
     *
     * @param category The category to bind.
     */
    fun bind(download: Download) {
        this.download = download
        // Update the chapter name.
        binding.chapterTitle.text = download.chapter.name

        // Update the manga title
        binding.mangaFullTitle.text = download.manga.title

        // Update the manga source
        binding.mangaSource.text = download.source.name

        // Update the progress bar and the number of downloaded pages
        val pages = download.pages
        if (pages == null) {
            binding.downloadProgress.progress = 0
            binding.downloadProgress.max = 1
            binding.downloadProgressText.text = ""
        } else {
            binding.downloadProgress.max = pages.size * 100
            notifyProgress()
            notifyDownloadedPages()
        }
    }

    /**
     * Updates the progress bar of the download.
     */
    fun notifyProgress() {
        val pages = download.pages ?: return
        if (binding.downloadProgress.max == 1) {
            binding.downloadProgress.max = pages.size * 100
        }
        binding.downloadProgress.setProgressCompat(download.totalProgress, true)
    }

    /**
     * Updates the text field of the number of downloaded pages.
     */
    fun notifyDownloadedPages() {
        val pages = download.pages ?: return
        binding.downloadProgressText.text = "${download.downloadedImages}/${pages.size}"
    }

    override fun onItemReleased(position: Int) {
        super.onItemReleased(position)
        adapter.downloadItemListener.onItemReleased(position)
        binding.container.isDragged = false
    }

    override fun onActionStateChanged(position: Int, actionState: Int) {
        super.onActionStateChanged(position, actionState)
        if (actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
            binding.container.isDragged = true
        }
    }

    private fun showPopupMenu(view: View) {
        view.popupMenu(
            menuRes = R.menu.download_single,
            initMenu = {
                findItem(R.id.move_to_top).isVisible = bindingAdapterPosition != 0
                findItem(R.id.move_to_bottom).isVisible =
                    bindingAdapterPosition != adapter.itemCount - 1
            },
            onMenuItemClick = {
                adapter.downloadItemListener.onMenuItemClick(bindingAdapterPosition, this)
            }
        )
    }
}
